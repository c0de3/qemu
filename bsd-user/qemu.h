/*
 *  qemu bsd user mode definition
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, see <http://www.gnu.org/licenses/>.
 */
#ifndef QEMU_H
#define QEMU_H

#include "qemu-user-common.h"
#include "exec/exec-all.h"

#undef DEBUG_REMAP
#ifdef DEBUG_REMAP
#endif /* DEBUG_REMAP */

#include "exec/user/abitypes.h"

enum BSDType {
    target_freebsd,
    target_netbsd,
    target_openbsd,
};
extern enum BSDType bsd_type;

#include "syscall_defs.h"
#include "target_syscall.h"
#include "target_signal.h"
#include "exec/gdbstub.h"

#if defined(CONFIG_USE_NPTL)
#define THREAD __thread
#else
#define THREAD
#endif

/* This struct is used to hold certain information about the image.
 * Basically, it replicates in user space what would be certain
 * task_struct fields in the kernel
 */
struct image_info {
    abi_ulong load_addr;
    abi_ulong start_code;
    abi_ulong end_code;
    abi_ulong start_data;
    abi_ulong end_data;
    abi_ulong start_brk;
    abi_ulong brk;
    abi_ulong start_mmap;
    abi_ulong mmap;
    abi_ulong rss;
    abi_ulong start_stack;
    abi_ulong entry;
    abi_ulong code_offset;
    abi_ulong data_offset;
    int       personality;
};

#define MAX_SIGQUEUE_SIZE 1024

struct sigqueue {
    struct sigqueue *next;
    //target_siginfo_t info;
};

struct emulated_sigtable {
    int pending; /* true if signal is pending */
    struct sigqueue *first;
    struct sigqueue info; /* in order to always have memory for the
                             first signal, we put it here */
};

/* NOTE: we force a big alignment so that the stack stored after is
   aligned too */
typedef struct TaskState {
    struct TaskState *next;
    int used; /* non zero if used */
    struct image_info *info;

    struct emulated_sigtable sigtab[TARGET_NSIG];
    struct sigqueue sigqueue_table[MAX_SIGQUEUE_SIZE]; /* siginfo queue */
    struct sigqueue *first_free; /* first free siginfo queue entry */
    int signal_pending; /* non zero if a signal may be pending */

    uint8_t stack[0];
} __attribute__((aligned(16))) TaskState;

void init_task_state(TaskState *ts);
extern const char *qemu_uname_release;
extern unsigned long mmap_min_addr;

/* ??? See if we can avoid exposing so much of the loader internals.  */
/*
 * MAX_ARG_PAGES defines the number of pages allocated for arguments
 * and envelope for the new program. 32 should suffice, this gives
 * a maximum env+arg of 128kB w/4KB pages!
 */
#define MAX_ARG_PAGES 32

/*
 * This structure is used to hold the arguments that are
 * used when loading binaries.
 */
struct linux_binprm {
        char buf[128];
        void *page[MAX_ARG_PAGES];
        abi_ulong p;
        int fd;
        int e_uid, e_gid;
        int argc, envc;
        char **argv;
        char **envp;
        char * filename;        /* Name of binary */
};

void do_init_thread(struct target_pt_regs *regs, struct image_info *infop);
abi_ulong loader_build_argptr(int envc, int argc, abi_ulong sp,
                              abi_ulong stringp, int push_ptr);
int loader_exec(const char * filename, char ** argv, char ** envp,
             struct target_pt_regs * regs, struct image_info *infop);

int load_elf_binary(struct linux_binprm * bprm, struct target_pt_regs * regs,
                    struct image_info * info);
int load_flt_binary(struct linux_binprm * bprm, struct target_pt_regs * regs,
                    struct image_info * info);

abi_long memcpy_to_target(abi_ulong dest, const void *src,
                          unsigned long len);
void target_set_brk(abi_ulong new_brk);
abi_long do_brk(abi_ulong new_brk);
void syscall_init(void);
abi_long do_freebsd_syscall(void *cpu_env, int num, abi_long arg1,
                            abi_long arg2, abi_long arg3, abi_long arg4,
                            abi_long arg5, abi_long arg6, abi_long arg7,
                            abi_long arg8);
abi_long do_netbsd_syscall(void *cpu_env, int num, abi_long arg1,
                           abi_long arg2, abi_long arg3, abi_long arg4,
                           abi_long arg5, abi_long arg6);
abi_long do_openbsd_syscall(void *cpu_env, int num, abi_long arg1,
                            abi_long arg2, abi_long arg3, abi_long arg4,
                            abi_long arg5, abi_long arg6);
void gemu_log(const char *fmt, ...) GCC_FMT_ATTR(1, 2);
extern THREAD CPUState *thread_cpu;
void cpu_loop(CPUArchState *env);
char *target_strerror(int err);
int get_osversion(void);
void fork_start(void);
void fork_end(int child);

#include "qemu/log.h"

/* strace.c */
struct syscallname {
    int nr;
    const char *name;
    const char *format;
    void (*call)(const struct syscallname *,
                 abi_long, abi_long, abi_long,
                 abi_long, abi_long, abi_long);
    void (*result)(const struct syscallname *, abi_long);
};

void
print_freebsd_syscall(int num,
                      abi_long arg1, abi_long arg2, abi_long arg3,
                      abi_long arg4, abi_long arg5, abi_long arg6);
void print_freebsd_syscall_ret(int num, abi_long ret);
void
print_netbsd_syscall(int num,
                     abi_long arg1, abi_long arg2, abi_long arg3,
                     abi_long arg4, abi_long arg5, abi_long arg6);
void print_netbsd_syscall_ret(int num, abi_long ret);
void
print_openbsd_syscall(int num,
                      abi_long arg1, abi_long arg2, abi_long arg3,
                      abi_long arg4, abi_long arg5, abi_long arg6);
void print_openbsd_syscall_ret(int num, abi_long ret);
extern int do_strace;

/* signal.c */
void process_pending_signals(CPUArchState *cpu_env);
void signal_init(void);
//int queue_signal(CPUArchState *env, int sig, target_siginfo_t *info);
//void host_to_target_siginfo(target_siginfo_t *tinfo, const siginfo_t *info);
//void target_to_host_siginfo(siginfo_t *info, const target_siginfo_t *tinfo);
long do_sigreturn(CPUArchState *env);
long do_rt_sigreturn(CPUArchState *env);
abi_long do_sigaltstack(abi_ulong uss_addr, abi_ulong uoss_addr, abi_ulong sp);

/* mmap.c */
int target_mprotect(abi_ulong start, abi_ulong len, int prot);
abi_long target_mmap(abi_ulong start, abi_ulong len, int prot,
                     int flags, int fd, abi_ulong offset);
int target_munmap(abi_ulong start, abi_ulong len);
abi_long target_mremap(abi_ulong old_addr, abi_ulong old_size,
                       abi_ulong new_size, unsigned long flags,
                       abi_ulong new_addr);
int target_msync(abi_ulong start, abi_ulong len, int flags);
extern unsigned long last_brk;
#if defined(CONFIG_USE_NPTL)
void mmap_fork_start(void);
void mmap_fork_end(int child);
#endif

/* main.c */
extern unsigned long x86_stack_size;

#if defined(CONFIG_USE_NPTL)
#include <pthread.h>
#endif

#endif /* QEMU_H */
