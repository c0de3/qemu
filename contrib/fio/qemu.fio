; Read 4 files with aio at different depths
[global]
ioengine=./qemu.so
direct=1
qemu_aio=native
create_on_open=1
rw=randread
bs=128k
filesize=128m
runtime=10s
time_based

[file1]
filename=file1
iodepth=4

[file2]
filename=file2
iodepth=32

[file3]
filename=file3
iodepth=8

[file4]
filename=file4
iodepth=16
