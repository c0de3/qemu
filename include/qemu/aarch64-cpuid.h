#ifndef QEMU_AARCH64_CPUID_H
#define QEMU_AARCH64_CPUID_H

#if defined (__aarch64__)
uint64_t get_aarch64_cpu_id(void);
bool is_thunderx_pass2_cpu(void);
#endif

#endif
