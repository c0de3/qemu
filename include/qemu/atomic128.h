/*
 * Simple interface for 128-bit atomic operations.
 *
 * Copyright (C) 2018 Linaro, Ltd.
 *
 * This work is licensed under the terms of the GNU GPL, version 2 or later.
 * See the COPYING file in the top-level directory.
 *
 * See docs/devel/atomics.txt for discussion about the guarantees each
 * atomic primitive is meant to provide.
 */

#ifndef QEMU_ATOMIC128_H
#define QEMU_ATOMIC128_H

/*
 * GCC is a house divided about supporting large atomic operations.
 *
 * For hosts that only have large compare-and-swap, a legalistic reading
 * of the C++ standard means that one cannot implement __atomic_read on
 * read-only memory, and thus all atomic operations must synchronize
 * through libatomic.
 *
 * See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80878
 *
 * This interpretation is not especially helpful for QEMU.
 * For softmmu, all RAM is always read/write from the hypervisor.
 * For user-only, if the guest doesn't implement such an __atomic_read
 * then the host need not worry about it either.
 *
 * Moreover, using libatomic is not an option, because its interface is
 * built for std::atomic<T>, and requires that *all* accesses to such an
 * object go through the library.  In our case we do not have an object
 * in the C/C++ sense, but a view of memory as seen by the guest.
 * The guest may issue a large atomic operation and then access those
 * pieces using word-sized accesses.  From the hypervisor, we have no
 * way to connect those two actions.
 *
 * Therefore, special case each platform.
 */

#if defined(CONFIG_ATOMIC128)
static inline Int128 atomic16_cmpxchg(Int128 *ptr, Int128 cmp, Int128 new)
{
    return atomic_cmpxchg__nocheck(ptr, cmp, new);
}
# define HAVE_CMPXCHG128 1
#elif defined(CONFIG_CMPXCHG128)
static inline Int128 atomic16_cmpxchg(Int128 *ptr, Int128 cmp, Int128 new)
{
    return __sync_val_compare_and_swap_16(ptr, cmp, new);
}
# define HAVE_CMPXCHG128 1
#elif defined(__aarch64__)
/* Through gcc 8, aarch64 has no support for 128-bit at all.  */
static inline Int128 atomic16_cmpxchg(Int128 *ptr, Int128 cmp, Int128 new)
{
    uint64_t cmpl = cmp, cmph = cmp >> 64;
    uint64_t newl = new, newh = new >> 64;
    uint64_t oldl, oldh;
    uint32_t tmp;

    asm("0: ldaxp %[oldl], %[oldh], %[mem]\n\t"
        "cmp %[oldl], %[cmpl]\n\t"
        "ccmp %[oldh], %[cmph], #0, eq\n\t"
        "b.ne 1f\n\t"
        "stlxp %w[tmp], %[newl], %[newh], %[mem]\n\t"
        "cbz %w[tmp], 0b\n"
        "1:"
        : [mem] "+m"(*ptr), [tmp] "=&r"(tmp),
          [oldl] "=&r"(oldl), [oldh] "=r"(oldh)
        : [cmpl] "r"(cmpl), [cmph] "r"(cmph),
          [newl] "r"(newl), [newh] "r"(newh)
        : "memory", "cc");

    return int128_make128(oldl, oldh);
}
# define HAVE_CMPXCHG128 1
#endif /* Some definition for HAVE_CMPXCHG128 */


#if defined(CONFIG_ATOMIC128)
static inline Int128 atomic16_read(Int128 *ptr)
{
    return atomic_read__nocheck(ptr);
}

static inline void atomic16_set(Int128 *ptr, Int128 val)
{
    atomic_set__nocheck(ptr, val);
}

# define HAVE_ATOMIC128 1
#elif !defined(CONFIG_USER_ONLY)
# ifdef __aarch64__
/* We can do better than cmpxchg for AArch64.  */
static inline Int128 atomic16_read(Int128 *ptr)
{
    uint64_t l, h;
    uint32_t tmp;

    /* The load must be paired with the store to guarantee not tearing.  */
    asm("0: ldxp %[l], %[h], %[mem]\n\t"
        "stxp %w[tmp], %[l], %[h], %[mem]\n\t"
        "cbz %w[tmp], 0b"
        : [mem] "+m"(*ptr), [tmp] "=r"(tmp), [l] "=r"(l), [h] "=r"(h));

    return int128_make128(l, h);
}

static inline void atomic16_set(Int128 *ptr, Int128 val)
{
    uint64_t l = val, h = val >> 64, t1, t2;

    /* Load into temporaries to acquire the exclusive access lock.  */
    asm("0: ldxp %[t1], %[t2], %[mem]\n\t"
        "stxp %w[t1], %[l], %[h], %[mem]\n\t"
        "cbz %w[t1], 0b"
        : [mem] "+m"(*ptr), [t1] "=&r"(t1), [t2] "=&r"(t2)
        : [l] "r"(l), [h] "r"(h));
}

#  define HAVE_ATOMIC128 1
# elif HAVE_CMPXCHG128
static inline Int128 atomic16_read(Int128 *ptr)
{
    /* Maybe replace 0 with 0, returning the old value.  */
    return atomic16_cmpxchg(ptr, 0, 0);
}

static inline void atomic16_set(Int128 *ptr, Int128 val)
{
    Int128 old = *ptr, cmp;
    do {
        cmp = old;
        old = atomic16_cmpxchg(ptr, cmp, val);
    } while (old != cmp);
}

#  define HAVE_ATOMIC128 1
# endif
#endif

/*
 * Fallback definitions that must be optimized away, or error.
 */

#ifndef HAVE_CMPXCHG128
Int128 __attribute__((error("unsupported cmpxchg")))
    atomic16_cmpxchg(Int128 *ptr, Int128 cmp, Int128 new);
# define HAVE_CMPXCHG128 0
#endif

#ifndef HAVE_ATOMIC128
Int128 __attribute__((error("unsupported atomic16_read")))
    atomic16_read(Int128 *ptr, Int128 cmp, Int128 new);
Int128 __attribute__((error("unsupported atomic16_set")))
    atomic16_set(Int128 *ptr, Int128 cmp, Int128 new);
# define HAVE_ATOMIC128 0
#endif

#endif /* QEMU_ATOMIC128_H */
