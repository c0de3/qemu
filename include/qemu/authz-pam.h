/*
 * QEMU pam authorization driver
 *
 * Copyright (c) 2016 Red Hat, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see <http://www.gnu.org/licenses/>.
 *
 */

#ifndef QAUTHZ_PAM_H__
#define QAUTHZ_PAM_H__

#include "qemu/authz.h"


#define TYPE_QAUTHZ_PAM "authz-pam"

#define QAUTHZ_PAM_CLASS(klass) \
     OBJECT_CLASS_CHECK(QAuthZPamClass, (klass), \
                        TYPE_QAUTHZ_PAM)
#define QAUTHZ_PAM_GET_CLASS(obj) \
     OBJECT_GET_CLASS(QAuthZPamClass, (obj), \
                      TYPE_QAUTHZ_PAM)
#define QAUTHZ_PAM(obj) \
     INTERFACE_CHECK(QAuthZPam, (obj), \
                     TYPE_QAUTHZ_PAM)

typedef struct QAuthZPam QAuthZPam;
typedef struct QAuthZPamClass QAuthZPamClass;


/**
 * QAuthZPam:
 *
 * This authorization driver provides a pam mechanism
 * for granting access by matching user names against a
 * list of globs. Each match rule has an associated policy
 * and a catch all policy applies if no rule matches
 *
 * To create an instance of this class via QMP:
 *
 *  {
 *    "execute": "object-add",
 *    "arguments": {
 *      "qom-type": "authz-pam",
 *      "id": "auth0",
 *      "parameters": {
 *        "rules": [
 *           { "match": "fred", "policy": "allow", "format": "exact" },
 *           { "match": "bob", "policy": "allow", "format": "exact" },
 *           { "match": "danb", "policy": "deny", "format": "exact" },
 *           { "match": "dan*", "policy": "allow", "format": "glob" }
 *        ],
 *        "policy": "deny"
 *      }
 *    }
 *  }
 *
 * Or via the CLI:
 *
 *   $QEMU                                                         \
 *    -object authz-pam,id=acl0,policy=deny,                    \
 *            match.0.name=fred,match.0.policy=allow,format=exact, \
 *            match.1.name=bob,match.1.policy=allow,format=exact,  \
 *            match.2.name=danb,match.2.policy=deny,format=exact,  \
 *            match.3.name=dan\*,match.3.policy=allow,format=exact
 *
 */
struct QAuthZPam {
    QAuthZ parent_obj;

    char *service;
};


struct QAuthZPamClass {
    QAuthZClass parent_class;
};


QAuthZPam *qauthz_pam_new(const char *id,
                          const char *service,
                          Error **errp);


#endif /* QAUTHZ_PAM_H__ */

