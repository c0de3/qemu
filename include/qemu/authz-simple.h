/*
 * QEMU simple authorization driver
 *
 * Copyright (c) 2016 Red Hat, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see <http://www.gnu.org/licenses/>.
 *
 */

#ifndef QAUTHZ_SIMPLE_H__
#define QAUTHZ_SIMPLE_H__

#include "qemu/authz.h"


#define TYPE_QAUTHZ_SIMPLE "authz-simple"

#define QAUTHZ_SIMPLE_CLASS(klass) \
     OBJECT_CLASS_CHECK(QAuthZSimpleClass, (klass), \
                        TYPE_QAUTHZ_SIMPLE)
#define QAUTHZ_SIMPLE_GET_CLASS(obj) \
     OBJECT_GET_CLASS(QAuthZSimpleClass, (obj), \
                      TYPE_QAUTHZ_SIMPLE)
#define QAUTHZ_SIMPLE(obj) \
     INTERFACE_CHECK(QAuthZSimple, (obj), \
                     TYPE_QAUTHZ_SIMPLE)

typedef struct QAuthZSimple QAuthZSimple;
typedef struct QAuthZSimpleClass QAuthZSimpleClass;


/**
 * QAuthZSimple:
 *
 * This authorization driver provides a simple mechanism
 * for granting access by matching user names against a
 * list of globs. Each match rule has an associated policy
 * and a catch all policy applies if no rule matches
 *
 * To create an instance of this class via QMP:
 *
 *  {
 *    "execute": "object-add",
 *    "arguments": {
 *      "qom-type": "authz-simple",
 *      "id": "auth0",
 *      "parameters": {
 *        "rules": [
 *           { "match": "fred", "policy": "allow", "format": "exact" },
 *           { "match": "bob", "policy": "allow", "format": "exact" },
 *           { "match": "danb", "policy": "deny", "format": "exact" },
 *           { "match": "dan*", "policy": "allow", "format": "glob" }
 *        ],
 *        "policy": "deny"
 *      }
 *    }
 *  }
 *
 * Or via the CLI:
 *
 *   $QEMU                                                         \
 *    -object authz-simple,id=acl0,policy=deny,                    \
 *            match.0.name=fred,match.0.policy=allow,format=exact, \
 *            match.1.name=bob,match.1.policy=allow,format=exact,  \
 *            match.2.name=danb,match.2.policy=deny,format=exact,  \
 *            match.3.name=dan\*,match.3.policy=allow,format=exact
 *
 */
struct QAuthZSimple {
    QAuthZ parent_obj;

    QAuthZSimplePolicy policy;
    QAuthZSimpleRuleList *rules;
};


struct QAuthZSimpleClass {
    QAuthZClass parent_class;
};


QAuthZSimple *qauthz_simple_new(const char *id,
                                QAuthZSimplePolicy policy,
                                Error **errp);

ssize_t qauthz_simple_append_rule(QAuthZSimple *auth,
                                  const char *match,
                                  QAuthZSimplePolicy policy,
                                  QAuthZSimpleFormat format,
                                  Error **errp);

ssize_t qauthz_simple_insert_rule(QAuthZSimple *auth,
                                  const char *match,
                                  QAuthZSimplePolicy policy,
                                  QAuthZSimpleFormat format,
                                  size_t index,
                                  Error **errp);

ssize_t qauthz_simple_delete_rule(QAuthZSimple *auth,
                                  const char *match);


#endif /* QAUTHZ_SIMPLE_H__ */

