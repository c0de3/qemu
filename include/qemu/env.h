/*
 * Copyright (c) 2020 Oracle and/or its affiliates.
 *
 * This work is licensed under the terms of the GNU GPL, version 2.
 * See the COPYING file in the top-level directory.
 *
 */

#ifndef QEMU_ENV_H
#define QEMU_ENV_H

#define FD_PREFIX "QEMU_FD_"
#define ADDR_PREFIX "QEMU_ADDR_"
#define LEN_PREFIX "QEMU_LEN_"
#define BOOL_PREFIX "QEMU_BOOL_"

typedef int (*walkenv_cb)(const char *name, const char *val, void *handle);

bool getenv_ram(const char *name, void **addrp, size_t *lenp);
void setenv_ram(const char *name, void *addr, size_t len);
void unsetenv_ram(const char *name);
int getenv_fd(const char *name);
void setenv_fd(const char *name, int fd);
void unsetenv_fd(const char *name);
bool getenv_bool(const char *name);
void setenv_bool(const char *name, bool val);
void unsetenv_bool(const char *name);
int walkenv(const char *prefix, walkenv_cb cb, void *handle);
void printenv(void);

#endif
