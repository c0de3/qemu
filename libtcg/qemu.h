#ifndef QEMU_H
#define QEMU_H

#include "qemu-user-common.h"

#endif /* QEMU_H */
