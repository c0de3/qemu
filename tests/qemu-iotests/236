#!/usr/bin/env python
#
# Test bitmap merges.
#
# Copyright (c) 2018 John Snow for Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# owner=jsnow@redhat.com

import sys
import os
import iotests
from iotests import log
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..', 'scripts'))
from qemu import QEMUMachine

iotests.verify_image_format(supported_fmts=['qcow2'])

patterns = [("0x5d", "0",         "64k"),
            ("0xd5", "1M",        "64k"),
            ("0xdc", "32M",       "64k"),
            ("0xcd", "0x3ff0000", "64k")]  # 64M - 64K

overwrite = [("0xab", "0",         "64k"), # Full overwrite
             ("0xad", "0x00f8000", "64k"), # Partial-left (1M-32K)
             ("0x1d", "0x2008000", "64k"), # Partial-right (32M+32K)
             ("0xea", "0x3fe0000", "64k")] # Adjacent-left (64M - 128K)

with iotests.FilePath('img') as img_path, \
     iotests.VM() as vm:

    log('--- Preparing image & VM ---\n')
    iotests.qemu_img_pipe('create', '-f', iotests.imgfmt, img_path, '64M')
    vm.add_drive(img_path)
    vm.launch()

    log('--- Adding preliminary bitmaps A & B ---\n')
    vm.qmp_log("block-dirty-bitmap-add", node="drive0", name="bitmapA")
    vm.qmp_log("block-dirty-bitmap-add", node="drive0", name="bitmapB")

    # Dirties 4 clusters. count=262144
    log('')
    log('--- Emulating writes ---\n')
    for p in patterns:
        cmd = "write -P%s %s %s" % p
        log(cmd)
        log(vm.hmp_qemu_io("drive0", cmd))

    vm.qmp_log("query-block", indent=2,
               filters=[iotests.filter_generated_node_ids,
                        iotests.filter_testfiles])

    log('')
    log('--- Disabling B & Adding C ---\n')
    vm.qmp_log("transaction", actions=[
        { "type": "block-dirty-bitmap-disable",
          "data": { "node": "drive0", "name": "bitmapB" }},
        { "type": "block-dirty-bitmap-add",
          "data": { "node": "drive0", "name": "bitmapC" }}
    ])

    log('')
    log('--- Emulating further writes ---\n')
    # Dirties 6 clusters, 3 of which are new in contrast to "A".
    # A = 64 * 1024 * (4 + 3) = 458752
    # C = 64 * 1024 * 6       = 393216
    for p in overwrite:
        cmd = "write -P%s %s %s" % p
        log(cmd)
        log(vm.hmp_qemu_io("drive0", cmd))

    log('')
    log('--- Disabling A & C ---\n')
    vm.qmp_log("transaction", actions=[
        { "type": "block-dirty-bitmap-disable",
          "data": { "node": "drive0", "name": "bitmapA" }},
        { "type": "block-dirty-bitmap-disable",
          "data": { "node": "drive0", "name": "bitmapC" }}
    ])

    # A: 7 clusters
    # B: 4 clusters
    # C: 6 clusters
    vm.qmp_log("query-block", indent=2,
               filters=[iotests.filter_generated_node_ids,
                        iotests.filter_testfiles])

    log('')
    log('--- Creating D as a merge of B & C ---\n')
    # Good hygiene: create a disabled bitmap as a merge target.
    vm.qmp_log("transaction", actions=[
        { "type": "block-dirty-bitmap-add",
          "data": { "node": "drive0", "name": "bitmapD", "disabled": True }},
        { "type": "block-dirty-bitmap-merge",
          "data": { "node": "drive0", "target": "bitmapD",
                    "bitmaps": ["bitmapB", "bitmapC"] }}
    ])

    # A and D should now both have 7 clusters apiece.
    # B and C remain unchanged with 4 and 6 respectively.
    vm.qmp_log("query-block", indent=2,
               filters=[iotests.filter_generated_node_ids,
                        iotests.filter_testfiles])

    # A and D should be equivalent.
    vm.qmp_log('x-debug-block-dirty-bitmap-sha256',
               node="drive0", name="bitmapA")
    vm.qmp_log('x-debug-block-dirty-bitmap-sha256',
               node="drive0", name="bitmapD")

    vm.shutdown()
