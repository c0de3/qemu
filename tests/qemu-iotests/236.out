QA output created by 236

== Testing conversion to raw should success ==
*** done
